pimd/pim6_mroute.c

```diff
@@ -0,0 +1,1125 @@
+
+		if (!connected_src) {
+			if (PIM_DEBUG_IGMP_PACKETS) {
+				zlog_debug("Recv ICMPv6 packet on interface: %s from a non-connected source: %pI4",
+					   ifp->name, &ip6_hdr->ip6_src);
+			}
+			return 0;
+		}
+
+		pim_ifp = ifp->info;
+		IPV6_ADDR_COPY (&ifaddr, &connected_src->u.prefix6);
+		gm = pim_igmp_sock_lookup_ifaddr(pim_ifp->gm_socket_list, ifaddr);
+
+		if (PIM_DEBUG_IGMP_PACKETS) {
+			zlog_debug(
+				"%s(%s): icmpv6 kernel upcall on %s(%p) for %pI4 -> %pI4",
+				__func__, pim->vrf->name, ifp->name, gm,
+				&ip6_hdr->ip6_src, &ip6_hdr->ip6_dst);
+		}
+			//TODO parse mld packets here 
+		else if (PIM_DEBUG_IGMP_PACKETS) {
+			zlog_debug("No ICMPv6 socket on interface: %s with connected source: %pFX",
+				   ifp->name, connected_src);
+		}
+	} else if (ip6_hdr->ip6_nxt) {
+
+		next_hop = ip6_hdr->ip6_nxt;
+		while (!(got_icmpv6) || (buf_size - accessed_len >= 8)) {
+			switch (next_hop) {
+				case IPPROTO_ICMPV6:
+					ifp = if_lookup_by_index(ifindex, pim->vrf->vrf_id);
+
+					if (!ifp || !ifp->info)
+						return 0;
+
+					connected_src = pim_if_connected_to_source(ifp, ip6_hdr->ip6_src);
+
+					if (!connected_src) {
+						if (PIM_DEBUG_IGMP_PACKETS) {
+							zlog_debug("Recv ICMPv6 packet on interface: %s from a non-connected source: %pI4",
+									ifp->name, &ip6_hdr->ip6_src);
+						}
+						return 0;
+					}
+
+					pim_ifp = ifp->info;
+					IPV6_ADDR_COPY (&ifaddr, &connected_src->u.prefix6);
+					gm = pim_igmp_sock_lookup_ifaddr(pim_ifp->gm_socket_list, ifaddr);
+
+					if (PIM_DEBUG_IGMP_PACKETS) {
+						zlog_debug(
+								"%s(%s): icmpv6 kernel upcall on %s(%p) for %pI4 -> %pI4",
+								__func__, pim->vrf->name, ifp->name, gm,
+								&ip6_hdr->ip6_src, &ip6_hdr->ip6_dst);
+					}
+					//TODO parse mld packets here
+					got_icmpv6 = true;
+					break;
+				default:
+					ext_hdr_len = (((struct ip6_ext *)buf)->ip6e_len + 1) << 3;
+					next_hop = ((struct ip6_ext *)buf)->ip6e_nxt;
+					break;
+			}
+			buf += ext_hdr_len;
+			accessed_len += ext_hdr_len;
+		
+		}
+
+	} else {
+		msg = (const struct mrt6msg *)buf;
+
+		ifp = pim_if_find_by_mcast_if_index(pim, msg->im6_mif);
+
+		if (!ifp)
+			return 0;
+		if (PIM_DEBUG_MROUTE) {
+			zlog_debug(
+				"%s: pim kernel upcall %s type=%d ip_p=%d from fd=%d for (S,G)=(%pI4,%pI4) on %s mifi=%d  size=%d",
+				__func__, mrt6msgtype2str[msg->im6_msgtype],
+				msg->im6_msgtype, ip6_hdr->ip6_nxt,
+				pim->mroute_socket, &msg->im6_src, &msg->im6_dst, ifp->name,
+				msg->im6_mif, buf_size);
+		}
+
+		switch (msg->im6_msgtype) {
+		case MRT6MSG_WRONGMIF:
+			return pim_mroute_msg_wrongmif(pim->mroute_socket, ifp,
+						       msg);
+		case MRT6MSG_NOCACHE:
+			return pim_mroute_msg_nocache(pim->mroute_socket, ifp,
+						      msg);
+		case MRT6MSG_WHOLEPKT:
+			return pim_mroute_msg_wholepkt(pim->mroute_socket, ifp,
+						       (const char *)msg);
+		default:
+			break;
+		}
+	}
+
+	return 0;
+}
+
+static int mroute_read(struct thread *t)
```

Comment:
Same applies here. If the function is used only here, there is no need to keep the function name same as v4.